### PR TITLE
reactor_backend - Fix backend selector logic

### LINKED ISSUE #1058: reactor_backend - has_enough_aio_nr() always evaluate to true when smp::count is 0 and reactor detector fail early if aio nr is exhausted
In a situation when the running system concurrent AIO contexts is exhausted (eg: `aio-max-nr == aio-nr`), seastar will fail to fallback to epoll backend. 

This happens for two reasons:

`reactor_backend_selector::available` will first `detect_aio_poll()` and - only after it suceeds - it will evaluate whether the system `has_enough_aio_nr()`. This logic is incorrect, as we must first evaluate whether we first have enough AIO NRs before detecting the AIO poll, as AIO poll throws upon failure to invoke `io_setup()`.

Within `has_enough_aio_nr()`, the check has the following logic: `aio_max_nr - aio_nr < reactor::max_aio * smp::count` which one would naively assume is correct. However, as defined within [reactor.cc](https://github.com/scylladb/seastar/blob/master/src/core/reactor.cc#L3560), `smp::count` evaluates to 0 during initialization, which means that the function will always skip the conditional and - thus - seastar will fail to fallback to epoll backend.

This needs to be fixed, as when the system has no AIO contexts available even a simple `scylla --help` will throw:

```shell
$ cat /proc/sys/fs/aio-max-nr
0
$ scylla --help 
FATAL: Exception during startup, aborting: std::runtime_error (Could not setup Async I/O: Resource temporarily unavailable. The most common cause is not enough request capacity in /proc/sys/fs/aio-max-nr. Try increasing that number or reducing the amount of logical CPUs available for your application)
```

It is worth to mention that, later on - after initialization and parsing the relevant command line parameters - a new check is done in order to ensure the system has enough AIO contexts, under `smp::adjust_max_networking_aio_io_control_blocks` and under `aio_storage_context::aio_storage_context`, which means that trying to start Scylla with an exhausted amount of AIO NRs will fail anyway, but at least the user will be able to print `--help` rather than failing early. ;-)

### PR DESCRIPTION
**Pull request by fee-mendes**
In a situation when the running system concurrent AIO contexts is exhausted (eg: aio-max-nr == aio-nr), seastar will fail to fallback to epoll backend.

reactor_backend_selector::available will first detect_aio_poll() and - only after it suceeds - it will evaluate whether the system has_enough_aio_nr(). This logic is incorrect, as we must first evaluate whether we first have enough AIO NRs before detecting the AIO poll, as AIO poll throws upon failure to invoke io_setup().

Within has_enough_aio_nr(), the check has the following logic: aio_max_nr - aio_nr < reactor::max_aio * smp::count which one would naively assume is correct. However, as defined within reactor.cc, smp::count evaluates to 0 during initialization, which means that the function will always skip the conditional and - thus - seastar will fail to fallback to epoll backend.

This needs to be fixed, as when the system has no AIO contexts available even a simple scylla --help will throw.

Fixes #1058

### CODE REVIEW

File: src/core/reactor_backend.cc [hunk 1]
```diff
@@ -1146,7 +1146,13 @@ bool reactor_backend_selector::has_enough_aio_nr() {
      * So this method calculates:
      *  Available AIO on the system - (request AIO per-cpu * ncpus)
      */
-    if (aio_max_nr - aio_nr < reactor::max_aio * smp::count) {
+    unsigned _smp_local_count = 1;
```
Comment (vladzcloudius) [inline src/core/reactor_backend.cc:1149]: We don't prepend local variables names with `_`

File: src/core/reactor_backend.cc [hunk 2]
```diff
@@ -1146,7 +1146,13 @@ bool reactor_backend_selector::has_enough_aio_nr() {
...
+
+    if (smp::count > 0){
+       _smp_local_count = smp::count;
+    }
+
```
Comment (vladzcloudius) [inline src/core/reactor_backend.cc:1154]: I don't understand how the above is going to resolve the problem at hand.
I believe that the only acceptable value you need is a real (!!) smp::count, e.g. 64 on a machine with 64 CPUs. Having 1 instead as your code would do is not addressing the issue AFAIU.

Comment (vladzcloudius) [reply src/core/reactor_backend.cc:1154]: The proper fix IMO is to make the original code run when smp::count evaluates to the expected value.
Using smp::count before it's initialized is simply a bug.

**Review by vladzcloudius — CHANGES_REQUESTED** [review]

File: src/core/reactor_backend.cc (hunk 1 above)
Comment (fee-mendes) [reply src/core/reactor_backend.cc:1149]: Could you point me to the relevant documentation where this is specified? I took a quick glance at the code and there are local variable names starting with a `_`. No problem over changing the variable name as long as it is indeed a practice. :)

File: src/core/reactor_backend.cc (hunk 2 above)
Comment (fee-mendes) [reply src/core/reactor_backend.cc:1154]: This is somehow documented a couple lines above -- I think what was unexpected is that we initialize `smp::count = 0`. That's the relevant comment already within the code:

```cpp
    /* reactor_backend_selector::available() will be execute in early stage,
     * it's before io_setup() issued, and not per-cpu basis.
     * So this method calculates:
     *  Available AIO on the system - (request AIO per-cpu * ncpus)
     */
```

TLDR; the backend selection happens early before we eventually parse the user command-line at `smp::configure`, so that means we initialize `smp::count = 0` and this is what causes the validation error. 

So that gives us two options:

1. Stick with the proposed solution here or,
2. Initialize `smp::count = 1` as it makes no sense - indeed - to start it with 0 (we can't run seastar in a pencil yet AFAIU)

Let me know which option seems better and I will push it with the relevant changes.

Comment (vladzcloudius) [reply src/core/reactor_backend.cc:1154]: @xemul Any comments?

File: src/core/reactor_backend.cc (hunk 1 above)
Comment (xemul) [reply src/core/reactor_backend.cc:1149]: It's indeed a practice. `_` usually comes as a prefix for private class members, not for class variables unless there's a _really_ _good_ reason for it.

**Comment by xemul** [comment]
@fee-mendes , it looks like there are two problems here. First with incorrect `smp::count` usage and the other one with swapped AIO detection/setup calls. Please, split the patch into two.

**Comment by fee-mendes** [comment]
Closed as I mistakenly merged upstream to my branch :-D 

Follow-up at https://github.com/scylladb/seastar/pull/1060 / https://github.com/scylladb/seastar/pull/1061
